fix: Promise pollyfil loaded before webpack bundle

rkostrzewski opened the pull request:
Closes #101.

Tested in Chrome on desktop & mobile and IE 11. Chrome doesn't complain about document.write (ofc I've used `typeof fetch === "function"`). Rendering is not blocked in IE. :smile:

lukeed commented on src/lib/webpack-config.js:
```diff
@@ -433,6 +435,7 @@ const production = config => addPlugins([
 		minify: true,
 		stripPrefix: config.cwd,
 		staticFileGlobsIgnorePatterns: [
+			/polyfills\.js$/,
```
Did you ignore this because all `serviceWorker`-compatible browsers don't need a `fetch` or `Promise` polyfill?

_(^ I'm not certain that that's 100% true)_

rkostrzewski replied:
From what I understand `serviceWorker` implementation requires `fetch` implementation which in turn requires `Promise` implementation. Based on [FetchEvent](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent). I could be wrong tough.

lukeed replied:
Ah, right. I believe that's the case too 👍 You upgraded my 90% certainty to 95% 😉

lukeed commented on src/resources/template.html:
```diff
@@ -24,5 +24,6 @@
 		<%= htmlWebpackPlugin.options.ssr({
 			url: '/'
 		}) %>
+		<script>"undefined"==typeof fetch&&document.write('<script src="<%= htmlWebpackPlugin.files.publicPath + "polyfills.js" %>"><\/script>')</script>
```
How about `(window.fetch===void 0) && ...`?

rkostrzewski replied:
Sounds great! I'll change it to `window.fetch==void 0&&document.write(...)` (5 less characters & html webpack plugin doesn't minify inline script 😢 )

lukeed replied:
Lol okay. I was just going for perf since `!== void 0` is faster than `typeof === 'undefined'`

developit replied:
if we're changing, maybe just go with `window.fetch || document.write(..)`.

developit commented:
Randomly commented in my review - do you guys think we could do something like this to work around using a blocking `document.write(<script>)`?

```html
<body>
  <script src="/bundle.js" defer></script>
  <script>window.fetch || document.body.appendChild(document.createElement('script')).src='/polyfills.js';</script>
</body>
```

ethanroday commented:
@developit Why not split out `Promise` from the rest of the polyfills and then take the approach you mentioned?
```html
<script>window.Promise || document.body.appendChild(....</script>
```
Then the rest of the polyfills could just be loaded as normal from webpack.

thangngoc89 commented:
@ethanroday currently, preact-cli only includes Promise and fetch. And you need promise supports for fetch anyway. So no splitting needed

lukeed commented:
Doesn't it _have_ to be blocking in order to ensure it loads before `bundle`?

developit commented:
@lukeed I wasn't sure if the async script (via createElement) would defer the `defer` script since those wait for document completion.

rkostrzewski commented:
@developit so I've done my research. 

`document.body.appendChild(document.createElement('script'))` doesn't block deferred scripts (kinda makes sense because `defer` only works for initial html, which `document.write` kinda mutates) so `document.write` has to stay. Good thing is `document.write(<script>)` does not block render (tested in Chrome & IE 11).

I've changed `async` to `defer` to preload `bundle.js` and added parsing blocking `document.write`. Tested in all IE >=9, latest Edge, Chrome, Firefox and android 4 browser. All is well :smile:

lukeed commented:
@rkostrzewski Makes sense, that's what I had "remembered" too.

However, I'm worried that changing from `async` to `defer` will hurt our load time. It shouldn't matter which attribute the main bundle is using since `document.write` is first & blocks, right?

rkostrzewski commented:
@lukeed perf should be pretty much the same for `async` & `defer` when using only one script. Additionally, `<script>` tag doesn't block `<script async>` as they are executed as soon as they are downloaded and stack is empty. `bundle.js` is much smaller than `polyfills.js` for initial project which means that it will most likely be downloaded earlier and executed earlier <- Promise doesn't exist yet - `require.ensure` & `import()` fail.

tl;dr `async` is random 😛

lukeed commented:
Haha okay 😆 Thanks!

rkostrzewski commented on src/resources/template.html:
```diff
@@ -11,6 +11,7 @@
 		<% for (var chunk of webpack.chunks) { %>
+		 <% if (chunk.names.length === 1 && chunk.names[0] === 'polyfills') continue; %>
 			<% for (var file of chunk.files) { %>
 				<% if (htmlWebpackPlugin.options.preload && file.match(/\.(js|css)$/)) { %>
```
BTW `htmlWebpackPlugin.options.preload` comes from `argv.preload` which is a hidden switch that can't be used. Gonna check whole codebase for hidden switches sometime soon.

rkostrzewski commented:
@lukeed gonna fix regex coz I just noticed it is wrong (`[bundle|polyfills]` should have been `(bundle|polyfills)`. Also `package-json.json` should not been comitted I guess.

reznord commented:
@rkostrzewski you are using npm5.x? I thought it wasn't that stable and it has many issues.

rkostrzewski commented:
@reznord works good enough for me. Hadn't had any issues yet.

reznord commented:
I was had issues with it when I was working with react-native, so thought of upgrading later.

reznord commented:
any more issues which I can take over?

rkostrzewski commented:
@reznord if you want feel free take anything with `help wanted` label :smile:
